core/pkg/artifacts/downloader.go

```diff
@@ -153,38 +154,56 @@ func (ad *ArtifactDownloader) downloadFiles(artifactID string, manifest Manifest
 						numDone++
 						continue
 					}
-					task := &filetransfer.Task{
-						FileKind: filetransfer.RunFileKindArtifact,
-						Type:     filetransfer.DownloadTask,
-						Path:     downloadLocalPath,
-						Url:      *entry.DownloadURL,
+					if entry.Ref != nil {
+						task := &filetransfer.ReferenceArtifactDownloadTask{
+							FileKind:  filetransfer.RunFileKindArtifact,
+							Path:      downloadLocalPath,
+							Reference: *entry.Ref,
+							Digest:    entry.Digest,
+							Size:      entry.Size,
+						}
+						versionId, ok := entry.Extra["versionID"]
+						if ok {
+							task.VersionId = versionId
+						}
+
+						task.TaskCompletionCallback = filetransfer.TaskCompletionCallback{
+							CompletionCallback: func() { taskResultsChan <- TaskResult{downloadLocalPath, task.Err, *entry.LocalPath} },
```

Comment:
`*entry.LocalPath` will panic if the server returns a malformed response. Could we turn that into an error instead?

Reply:
as in do a safety check when dereferencing it and return that err in the `TaskResult`?

Reply:
I think we'd actually error out earlier on this line: `downloadLocalPath := filepath.Join(ad.DownloadRoot, *entry.LocalPath)`, so maybe I should do a safety check there instead